### ISSUE TITLE
AP/AR BMI predicted for all values of covariate within individuals

### ISSUE THREAD

**Issue by burrowsk** [mcanouil/eggla#50]
### Bug description

When predicting data using compute_apar() on LMM data including a cohort specific covariate, the predictions of BMI are performed on both values of the binary variable for example:

Source covariate is coded as 1 = clinic, 2 = questionnaire

Individual 1 is coded as source=1
Individual 2 is coded as source=2

Output from predicted_data object (https://m.canouil.fr/eggla/articles/adiposity-peak-rebound.html)

sex | egg_id | egg_source | egg_ageyears | egg_bmi | AP | AR
-- | -- | -- | -- | -- | -- | --
0 | "1" | 1 | 0.25 | 16.42 | FALSE | FALSE
0 | "1" | 1 | 0.3 | 16.72 | FALSE | FALSE
0 | "1" | 1 | 0.35 | 16.98 | FALSE | FALSE
0 | "1" | 1 | 0.4 | 17.20 | FALSE | FALSE
0 | "1" | 1 | 0.45 | 17.37 | FALSE | FALSE
0 | "1" | 1 | 0.5 | 17.51 | FALSE | FALSE
0 | "1" | 1 | 0.55 | 17.61 | FALSE | FALSE
0 | "1" | 1 | 0.6 | 17.68 | FALSE | FALSE
0 | "1" | 1 | 0.65 | 17.71 | FALSE | FALSE
0 | "1" | 1 | 0.7 | 17.73 | TRUE | FALSE

sex | egg_id | egg_source | egg_ageyears | egg_bmi | AP | AR
-- | -- | -- | -- | -- | -- | --
1 | "2" | 1 | 0.25 | 15.3911 | FALSE | TRUE
1 | "2" | 2 | 0.25 | 15.0745 | FALSE | FALSE
1 | "2" | 1 | 0.3 | 15.7286 | FALSE | FALSE
1 | "2" | 2 | 0.3 | 15.4051 | FALSE | FALSE
1 | "2" | 1 | 0.35 | 16.0235 | FALSE | FALSE
1 | "2" | 2 | 0.35 | 15.6939 | FALSE | FALSE
1 | "2" | 1 | 0.4 | 16.2767 | FALSE | FALSE
1 | "2" | 2 | 0.4 | 15.9419 | FALSE | FALSE
1 | "2" | 1 | 0.45 | 16.4898 | FALSE | FALSE
1 | "2" | 2 | 0.45 | 16.1506 | FALSE | FALSE

Note that AP and AR appear to be predicted for each age increment, holding source =1. So, there are no duplicate AP or ARs which is good. For the predictions, would we not want to predict based on the individuals value for source, i.e those with 1 predicted with 1 and those with 2 predicted with 2 rather than 1?

Thanks!

### `eggla` version output

```r
[1] ‘0.12.3’
```

### Checklist

- [X] [formatted your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read?
- [X] included a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] documented the version of `eggla` you're running, by pasting the output from running `packageVersion('eggla')` in the "`eggla` version output" text area?
- [X] documented which operating system you're running? If on Linux, please provide the specific distribution as well.

**Comment by mcanouil** [comment]
I need to make this function more robust, when retrieving individuals values.
A fix will be soon available.

**Comment by mcanouil** [comment]
@burrowsk are your IDs not unique, _i.e._, ID "1" can exist in "source 1" and "source 2"?

**Comment by burrowsk** [comment]
That's a really good point, no not  unique IDs i.e ID one can have several observations of height from clinic and several from questionnaire. So really it is a time variant covariate.

**Comment by mcanouil** [comment]
To confirm, in the following table, is "2" from source 1 the same "2" as from source 2?
In other word, does it mean an individual has two types of record used to compute is BMI trajectory?
If that's the case the "true" trajectory for "2" is somewhere between "source 1" and "source 2".

For me, an ID can't have two records for the same age.

sex | egg_id | egg_source | egg_ageyears | egg_bmi | AP | AR
-- | -- | -- | -- | -- | -- | --
1 | "2" | 1 | 0.25 | 15.3911 | FALSE | TRUE
1 | "2" | 2 | 0.25 | 15.0745 | FALSE | FALSE

**Comment by burrowsk** [comment]
So the underlying observed phenotype data can look like this (dummy data):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/epwkb/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/epwkb/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



ID | age | sex | height | weight | source_str | source
-- | -- | -- | -- | -- | -- | --
"2" | 0.12 | 1 | 57 | 4.3 | Clinic | 1
"2" | 0.79 | 1 | 70 | 10.2 | Clinic | 1
"2" | 2.19 | 1 | 89 | 12.3 | Clinic | 1
"2" | 3.50 | 1 | 96.5 | 14 | Clinic | 1
"2" | 5.50 | 1 | 111.7 | 19.3 | Questionnaire | 2
"2" | 5.83 | 1 | 112 | 18 | Questionnaire | 2
"2" | 6.58 | 1 | 115 | 21 | Questionnaire | 2
"2" | 7.42 | 1 | 122.2 | 22.8 | Clinic | 1
"2" | 8.50 | 1 | 128.2 |   | Clinic | 1
"2" | 9.92 | 1 | 135.2 | 29.4 | Clinic | 1
"2" | 9.92 | 1 | 135 | 29 | Questionnaire | 2
"2" | 10.83 | 1 | 141.6 | 33.7 | Clinic | 1
"2" | 11.83 | 1 | 147.4 | 36.9 | Clinic | 1
"2" | 12.92 | 1 | 155 | 41.7 | Questionnaire | 1
"2" | 13.25 | 1 | 157.8 | 45 | Clinic | 2
"2" | 14.08 | 1 | 160.4 | 50.6 | Clinic | 1
"2" | 15.50 | 1 | 161.8 | 56.8 | Clinic | 1
"2" | 17.58 | 1 | 163 | 59 | Questionnaire | 2
"2" | 17.83 | 1 | 162.5 | 57.8 | Clinic | 1



</body>

</html>



Where there is a clinic and questionnaire entry for the same age, these are removed as same day during Daymont cleaning. The cleaned data are then entered into the model like this, any observations whether clinic or questionnaire are entered together for each individual and "source" is added as a fixed effect covariate. 

A previous example of source as fixed effect is in https://journals.sagepub.com/doi/10.1177/0962280213503925?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed

Though I am now confused as to how this works with predictions, should we predict BMI based on the age intervals and source as 1?

**Comment by mcanouil** [comment]
It seems we should predict independently of source in that case, or any other covariate. 🤔
I think I have a better understanding of the context.
We can discuss it and I'll see what can be done later today.

**Comment by mcanouil** [comment]
@burrowsk Any update from your side?
-> I am going to implement a geometric mean estimates of the AP/AR on predicted values.

**Comment by burrowsk** [comment]
> @burrowsk Any update from your side? -> I am going to implement a geometric mean estimates of the AP/AR on predicted values.

Not right now, Kate is away until 21st. I did mention this to Nicole, her instinct was to predict based on source = clinic (1) as most other cohorts are clinic data (this is also the most abundant measure in ALSPAC (~70% of measures).

Came across this, though not yet looked into whether this is something useful for us https://rdrr.io/cran/LMMstar/man/predict.lmm.html

**Comment by mcanouil** [comment]
Ok, I'll figure something that will allow to specify something for you but stays general.

**Comment by burrowsk** [comment]
@mcanouil am I right in that the only outcomes we are using stats::predict() (eggla::bmi_predict()) for is to compute BMI/age of AP/AR? Nothing else is affected by this?

**Comment by mcanouil** [comment]
That is correct, I am currently working on this and will publish a branch today with an implementation to allow filtering or aggregation.

**Comment by burrowsk** [comment]
That’s awesome, thanks!!

------------------------------
Dr. Kimberley Burrows
Senior Research Associate in Epidemiology
MRC Integrative Epidemiology Unit
Population Health Sciences
Bristol Medical School
University of Bristol
Room OS1
Oakfield House, Oakfield Grove, Bristol, BS8 2BN

Email: ***@***.***

http://www.bristol.ac.uk/integrative-epidemiology/
Profile-Kimberley-Burrows<https://research-information.bris.ac.uk/en/persons/kimberley-burrows(2513bf22-07b2-44e9-ad72-133c3be48c17).html>

A top 3 UK university with leading employers (High Fliers 2017)
A top 5 UK university for research (REF 2014)
A world top 50 university (QS Ranking 2016)

My pronouns are she/her

**Comment by burrowsk** [comment]
I had some feedback from Kate who said:
"if the covariates have just a constant effect over time (e.g. the clinic/self-report contrast just has one estimate) then it won’t actually matter but I would pick clinic as that likely to be most accurate. So predict what you would have got had it been a clinic measure."

**Comment by mcanouil** [comment]
@burrowsk Thanks for this update.
I implemented two things in #70 
- A warning is displayed when multiple BMI measures are detected after BMI prediction and BMI values are aggretated using the geometric mean.
  ```
  Warning message:
  Multiple BMI measures (for the same age) have been detected and are aggregated using geometric mean!
  Use "filter" parameter to apply some filtering, e.g., filter = "source == 'clinic'". 
  ```
- A new `filter` parameter (called `filter_apar` in `run_eggla_lmm()`), allowing you to define any filter to be applied in "i" of a `data.table  object.
  ``` r
  data("bmigrowth")
  res <- egg_model(
    formula = log(bmi) ~ age,
    data = bmigrowth[bmigrowth[["sex"]] == 0, ],
    id_var = "ID",
    random_complexity = 1
  )
  #> Fitting model:
  #>   nlme::lme(
  #>     fixed = log(bmi) ~ gsp(age, knots = c(1, 8, 12), degree = rep(3, 4), smooth = rep(2, 3)),
  #>     data = data,
  #>     random = ~ gsp(age, knots = c(1, 8, 12), degree = rep(1, 4), smooth = rep(2, 3)) | ID,
  #>     na.action = stats::na.omit,
  #>     method = "ML",
  #>     control = nlme::lmeControl(opt = "optim", maxIter = 500, msMaxIter = 500)
  #>   )
  predict_bmi(res)[]
  #>       egg_id egg_ageyears  egg_bmi
  #>    1:    082         0.25 16.21582
  #>    2:    082         0.30 16.57194
  #>    3:    082         0.35 16.88076
  #>    4:    082         0.40 17.14348
  #>    5:    082         0.45 17.36196
  #>   ---                             
  #> 9796:    068         9.80 30.49370
  #> 9797:    068         9.85 30.60051
  #> 9798:    068         9.90 30.70758
  #> 9799:    068         9.95 30.81492
  #> 9800:    068        10.00 30.92255
  
  ## For multiple sources of measures or multiple measures at one age
  set.seed(1234)
  dta <- bmigrowth[bmigrowth[["sex"]] == 0, ]
  dta[["source"]] <- c("A", "B")[rbinom(n = nrow(dta), size = 1, prob = 0.65) + 1]
  res <- egg_model(
    formula = log(bmi) ~ age + source,
    data = dta,
    id_var = "ID",
    random_complexity = 1
  )
  #> Fitting model:
  #>   nlme::lme(
  #>     fixed = log(bmi) ~ gsp(age, knots = c(1, 8, 12), degree = rep(3, 4), smooth = rep(2, 3)) + source,
  #>     data = data,
  #>     random = ~ gsp(age, knots = c(1, 8, 12), degree = rep(1, 4), smooth = rep(2, 3)) | ID,
  #>     na.action = stats::na.omit,
  #>     method = "ML",
  #>     control = nlme::lmeControl(opt = "optim", maxIter = 500, msMaxIter = 500)
  #>   )
  predict_bmi(res)[order(egg_id, egg_ageyears)]
  #> Warning in predict_bmi(res): Multiple BMI measures (for the same age) have been detected and are aggregated using geometric mean!
  #> Use "filter" parameter to apply some filtering, e.g., filter = "source == 'clinic'".
  #>       egg_id egg_ageyears  egg_bmi
  #>    1:    001         0.25 18.42096
  #>    2:    001         0.30 18.82627
  #>    3:    001         0.35 19.17792
  #>    4:    001         0.40 19.47726
  #>    5:    001         0.45 19.72642
  #>   ---                             
  #> 9796:    099         9.80 33.31745
  #> 9797:    099         9.85 33.44629
  #> 9798:    099         9.90 33.57549
  #> 9799:    099         9.95 33.70508
  #> 9800:    099        10.00 33.83508
  predict_bmi(res, filter = "source == 'A'")[order(egg_id, egg_ageyears)]
  #>       egg_id egg_source egg_ageyears  egg_bmi
  #>    1:    001          A         0.25 18.39947
  #>    2:    001          A         0.30 18.80430
  #>    3:    001          A         0.35 19.15554
  #>    4:    001          A         0.40 19.45454
  #>    5:    001          A         0.45 19.70340
  #>   ---                                        
  #> 9796:    099          A         9.80 33.27858
  #> 9797:    099          A         9.85 33.40727
  #> 9798:    099          A         9.90 33.53632
  #> 9799:    099          A         9.95 33.66576
  #> 9800:    099          A        10.00 33.79560
  ```
